### ISSUE TITLE
General discussion on the "services/caches/save_personal_note" method.

### ISSUE THREAD

**Issue by GoogleCodeExporter** [opencaching/okapi#302]
```
General discussion on the "services/caches/save_personal_note" method.

```

Original issue reported on code.google.com by `rygielski` on 25 Jan 2014 at 10:05

**Comment by pstorch** [comment]
Is there any size limit on `new_value`?
I was able to upload 4416873 bytes of Text. Didn't test any further ;)

**Comment by wrygiel** [comment]
This depends on OC installation. Whatever you upload, OKAPI will attempt to insert into the appropriate database field. However, OKAPI doesn't check if there's any limit on the database field. If there is, then I guess that MySQL will silently truncate your string at certain length.

I just checked database OC structures and it seems that OCPL will truncate your data on 64kB, while OCDE will allow 16MB.

**Comment by pstorch** [comment]
The reason behind my question is a c:geo issue: https://github.com/cgeo/cgeo/issues/6591
So I guess if we implement a 64kB Limit for OKAPI based Caches, we are on the safe side?

**Comment by wrygiel** [comment]
Yes, I think so. BTW, you will probably need to take into account that some UTF8 characters can eat much more than one byte.

**Comment by pstorch** [comment]
@wrygiel good point

**Comment by wrygiel** [comment]
Depending on the charset used to store the data, *all* characters (even those from BMP) can eat more than one byte. As far as I can recall, MySQL uses 2 or 3 bytes per each character. So, it might be the safest to limit this to ~20k characters.

**Comment by wrygiel** [comment]
Added a link in cc7f9a77.

**Comment by hxdimpf** [comment]
I believe the service isn't working as intended. Here is my test:

- Startcondition, a cache with zero text in the cache note
- my first request: `https://www.opencaching.de/okapi/services/caches/save_personal_notes?cache_code=OC17BD&old_value=&new_value=Test`
- Response: `{"saved_value": "Test", "replaced": true}` ... which is absolutely what I had expected
-  Next request, assuming that an empty `old_value` **and** an empty `new_value` will actually delete what's in the database, effectively make the personal note **empty**: `https://www.opencaching.de/okapi/services/caches/save_personal_notes?cache_code=OC17BD&old_value=&new_value=`
- Response: `{"saved_value": "Test\n\n", "replaced": false}` ... first of all, the `\n\n` is totally unexpected, second, I had expected to get `{saved_value": "", "replaced":true}`
- Final test: knowing that `Test\n\n` is out there, trying to clear it: `https://www.opencaching.de/okapi/services/caches/save_personal_notes?cache_code=OC17BD&old_value=Test\n\n&new_value=`
- Result: `{ "saved_value": "Test\n\n", "replaced": false}` .. Remember, I had **exactly** passed the correct `old_value`

With all  that experience, I would highly recommend to implement a parameter or option that makes the service unconditionally write `new_value` completely replacing what was potentially there. It is within the discretion of the client application to know what it wants to write to the personal cache notes. I don't  know of any use case where one wants to add text to what's already  out there. This is not an idempotent operation.

**Comment by wrygiel** [comment]
> Next request, assuming that an empty `old_value` **and** an empty `new_value` will actually delete what's in the database

It should. But in your case you have just set the value to "Test", so it's no longer empty. I'd try `old_value=Test&new_value=` instead.

> The `\n\n` is totally unexpected

It's not explicitly documented, but appending is done with a `\n\n` separator. So, if the client doesn't have access to `old_value`, then `<old_value>\n\n<new_value>` will be stored (two separate messages).

> * Final test: knowing that `Test\n\n` is out there, trying to clear it

Newlines are formatted as `\n` in JSON, but the input variables (the ones you use in the URL) do not accept JSON encoding. Try URL-encoding your values before using them in inputs.


: `https://www.opencaching.de/okapi/services/caches/save_personal_notes?cache_code=OC17BD&old_value=Test\n\n&new_value=`
> * Result: `{ "saved_value": "Test\n\n", "replaced": false}` .. Remember, I had **exactly** passed the correct `old_value`
> 
> With all that experience, I would highly recommend to implement a parameter or option that makes the service unconditionally write `new_value` completely replacing what was potentially there. It is within the discretion of the client application to know what it wants to write to the personal cache notes. I don't know of any use case where one wants to add text to what's already out there. This is not an idempotent operation.

**Comment by hxdimpf** [comment]
Ok, it is what it is. "works as designed". I can live with it but I have a hard time to understand the rationales for designing it that way. I'd want an "idempotent setter", for instance a semantics like this:

`myPCN = getPersonalCacheNote(cache_code)`

and an idempotent operation:

`setPersonalCacheNote(cache_code, myPCN)`

**Comment by wrygiel** [comment]
This is a simplified form of Optimistic Concurrency Control (OCC). It helps prevent unintentional data overrides by ensuring you've fetched the latest data before making an update. To understand OCC in depth, read the [Wikipedia article](https://en.wikipedia.org/wiki/Optimistic_concurrency_control).

**Comment by hxdimpf** [comment]
> This is a simplified form of Optimistic Concurrency Control (OCC). It helps prevent unintentional data overrides by ensuring you've fetched the latest data before making an update. To understand OCC in depth, read the [Wikipedia article](https://en.wikipedia.org/wiki/Optimistic_concurrency_control).

Interesting. Haven't heard of that before. I would understand this for extremely critical, perhaps financial transactions but in the given scenario, I'd consider that "overkill". But ok, I have my cicrumventions now in place.

Thanks for explaining the OCC.
